vm_eval.c

```diff
@@ -1958,7 +1958,14 @@ yield_under(VALUE under, VALUE self, int argc, const VALUE *argv, int kw_splat)
 	VM_FORCE_WRITE_SPECIAL_CONST(&VM_CF_LEP(ec->cfp)[VM_ENV_DATA_INDEX_SPECVAL], new_block_handler);
     }
 
+    //rb_obj_info_dump(under);
+    // Make crefs log that this is a special lazy singleton?
+
     cref = vm_cref_push(ec, under, ep, TRUE);
```

Comment:
It might be good to add an assertion `VM_ASSERT(singleton ? under == CLASS_OF(self) : under == self)`. Personally I like to remove `under` argument and pass `CLASS_OF(self)` to `vm_cref_push` when `singleton` is true, but I don't know whether @ko1 likes.